lib/posix/options/mmap.c

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2024, Tenstorrent AI ULC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <stddef.h>
+#include <sys/types.h>
+
+#include <kernel_arch_interface.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/fdtable.h>
+#include <zephyr/posix/sys/mman.h>
+#include <zephyr/posix/unistd.h>
+
+#define _page_size COND_CODE_1(CONFIG_MMU, (CONFIG_MMU_PAGE_SIZE), (PAGE_SIZE))
+
+int zvfs_ioctl(int fd, int cmd, va_list args);
+
+static int p2z(int prot, int pflags)
+{
+	bool rw = (prot & PROT_WRITE) != 0;
+	bool ex = (prot & PROT_EXEC) != 0;
+	bool fixed = (pflags & MAP_FIXED) != 0;
+	bool shared = (pflags & MAP_SHARED) != 0;
+	bool private = (pflags & MAP_PRIVATE) != 0;
+
+	if (!(shared ^ private)) {
+		return -1;
+	}
+
+	return (rw * K_MEM_PERM_RW) | (ex * K_MEM_PERM_EXEC) | (fixed * K_MEM_DIRECT_MAP);
+}
+
+static inline int zvfs_ioctl_wrap(int fd, int cmd, ...)
+{
+	int ret;
+	va_list args;
+
+	va_start(args, cmd);
+	ret = zvfs_ioctl(fd, cmd, args);
+	va_end(args);
+
+	return ret;
+}
+
+void *mmap(void *addr, size_t len, int prot, int flags, int fd, off_t off)
+{
+	int zflags = p2z(prot, flags);
+
+	if ((len == 0) || (zflags == -1)) {
+		errno = EINVAL;
+		return MAP_FAILED;
+	}
+
+	if (fd > 0) {
+		/* non-anonymous mapping */
+		if (zvfs_ioctl_wrap(fd, ZFD_IOCTL_MMAP, addr, len, prot, flags, off, &addr) < 0) {
+			return MAP_FAILED;
+		}
+
+		return addr;
+	}
+
+	if (!IS_ENABLED(CONFIG_MMU)) {
+		errno = ENOTSUP;
+		return MAP_FAILED;
+	}
+
+	/* anonymous mapping */
+	if ((flags & MAP_FIXED) == 0) {
+		addr = k_mem_map(len, zflags);
+	} else {
+		addr = k_mem_phys_map((uintptr_t)addr, ROUND_UP(len, _page_size), zflags);
+	}
+
+	if (addr == NULL) {
+		errno = ENOMEM;
+		return MAP_FAILED;
+	}
+
+	return addr;
+}
+
+int msync(void *addr, size_t length, int flags)
+{
+	ARG_UNUSED(addr);
+	ARG_UNUSED(length);
+	ARG_UNUSED(flags);
+
+	return 0;
```

Comment:
so this will just pass silently? should we return -1 (& ENOSYS) instead?

Reply:
Currently, it doesn't seem that there is a generic API to flush anonymously mapped pages. Also from what I understand, the memories used in this implementation so far use synchronized writes. I could double-check that though.

Eventually, when mapping regular files is implemented by the fs subsystem, there will need to be an ISR to catch writes, and those might also be buffered, but that's a future problem.

Otherwise, it may be useful to integrate the cache API eventually, but it's probably also premature at this point.

https://docs.zephyrproject.org/latest/hardware/cache/index.html

Reply:
Yeah, actually, so `k_mem_map_impl()` does not support explicit caching flags, so all writes will be synchronous. There is no need to fail, since doing nothing and succeeding has the expected result.

https://github.com/zephyrproject-rtos/zephyr/blob/453ab8a9a356acf475a965a777a370795effa255/kernel/mmu.c#L581